xrootd/client/session.go

```diff
@@ -0,0 +1,252 @@
+			if _, err := io.ReadFull(session.conn, resp.Data); err != nil {
+				if ctx.Err() != nil {
+					// something happened to the context.
+					// ignore this error.
+					continue
+				}
+				resp.Err = err
+			}
+
+			switch header.Status {
+			case xrdproto.Error:
+				resp.Err = header.Error(resp.Data)
+			case xrdproto.Wait:
+				go func(streamID xrdproto.StreamID) {
+					// TODO: fetch sleep value from response.
+					time.Sleep(time.Second * 5)
+					if _, err := session.conn.Write(session.requests[streamID]); err != nil {
+						log.Printf("received error: %v", err)
+					}
+				}(header.StreamID)
+				continue
+			case xrdproto.Redirect:
+				port := binary.BigEndian.Uint32(resp.Data)
+				parts := strings.Split(string(resp.Data[4:]), "?")
+				if len(parts) == 0 {
+					resp.Err = errors.Errorf("xrootd: could not parse redirect url %q", string(resp.Data[4:]))
+				} else {
+					opaque := ""
+					token := ""
+					if len(parts) > 1 {
+						opaque = parts[1]
+					}
+					if len(parts) > 2 {
+						token = parts[2]
+					}
+					resp.Redirection = &mux.Redirection{Host: parts[0], Port: int(port), Opaque: opaque, Token: token}
+				}
+			}
+
+			if err := session.mux.SendData(header.StreamID, resp); err != nil {
+				if ctx.Err() != nil {
+					// something happened to the context.
+					// ignore this error.
+					continue
+				}
+				panic(err)
+				// TODO: should we just ignore responses to unclaimed stream IDs?
+			}
+
+			if header.Status != xrdproto.OkSoFar {
+				session.mux.Unclaim(header.StreamID)
+			}
+		}
+	}
+}
+
+func (session *session) send(ctx context.Context, responseChannel mux.DataRecvChan, request []byte) ([]byte, *mux.Redirection, error) {
+	if _, err := session.conn.Write(request); err != nil {
+		return nil, nil, err
+	}
+
+	var data []byte
+
+	for {
+		select {
+		case resp, more := <-responseChannel:
+			if !more {
+				return data, nil, nil
+			}
+
+			if resp.Err != nil {
+				return nil, resp.Redirection, resp.Err
+			}
+
+			if resp.Redirection != nil {
+				return nil, resp.Redirection, nil
+			}
+
+			data = append(data, resp.Data...)
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+	panic("unreachable")
+}
+
+// Send sends the request to the server and stores the response inside the resp.
+func (session *session) Send(ctx context.Context, resp xrdproto.Response, req xrdproto.Request) (*mux.Redirection, error) {
+	streamID, responseChannel, err := session.mux.Claim()
+	if err != nil {
+		return nil, err
+	}
+
+	var wBuffer xrdenc.WBuffer
+	header := xrdproto.RequestHeader{streamID, req.ReqID()}
+	if err = header.MarshalXrd(&wBuffer); err != nil {
+		return nil, err
+	}
+	if err = req.MarshalXrd(&wBuffer); err != nil {
+		return nil, err
+	}
+	data := wBuffer.Bytes()
+
+	if session.signRequirements.Needed(req) {
+		data, err = session.sign(streamID, req.ReqID(), data)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	session.requests[streamID] = data
```

Comment:
I am unsure whether this won't trigger a data race (or a concurrent map-write, that will make the runtime panic (with Go>=1.10.x))
IIRC, only `session.send` is thread-safe.
couldn't this be handled inside `session.send` ?